drivers/sensor/adi/adxl362/adxl362_decoder.c

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright (c) 2024 Analog Devices Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "adxl362.h"
+#include <zephyr/sys/byteorder.h>
+
+#ifdef CONFIG_ADXL362_STREAM
+
+#define ADXL362_COMPLEMENT         0xf000
+
+static const uint32_t accel_period_ns[] = {
+	[ADXL362_ODR_12_5_HZ] = UINT32_C(10000000000) / 125,
+	[ADXL362_ODR_25_HZ] = UINT32_C(1000000000) / 25,
+	[ADXL362_ODR_50_HZ] = UINT32_C(1000000000) / 50,
+	[ADXL362_ODR_100_HZ] = UINT32_C(1000000000) / 100,
+	[ADXL362_ODR_200_HZ] = UINT32_C(1000000000) / 200,
+	[ADXL362_ODR_400_HZ] = UINT32_C(1000000000) / 400,
+};
+
+static const int32_t range_to_scale[] = {
+	/* See table 1 in specifications section of datasheet */
+	[ADXL362_RANGE_2G] = ADXL362_ACCEL_2G_LSB_PER_G,
+	[ADXL362_RANGE_4G] = ADXL362_ACCEL_4G_LSB_PER_G,
+	[ADXL362_RANGE_8G] = ADXL362_ACCEL_8G_LSB_PER_G,
+};
+
+static inline void adxl362_temp_convert_q31(q31_t *out, int16_t data_in)
+{
+/* See sensitivity and bias specifications in table 1 of datasheet */
+	data_in &= 0xFFF;
+
+	if (data_in & BIT(11)) {
+		data_in |= ADXL362_COMPLEMENT;
+	}
+
+	int64_t milli_c = (data_in - ADXL362_TEMP_BIAS_LSB) * ADXL362_TEMP_MC_PER_LSB +
+		      (ADXL362_TEMP_BIAS_TEST_CONDITION * 1000);
+
+	*out = CLAMP(((milli_c * 1000) + ((milli_c % 1000) * 1000)), INT32_MIN, INT32_MAX);
+}
+
+static inline void adxl362_accel_convert_q31(q31_t *out, int16_t data_in, int32_t range)
+{
+	data_in &= 0xFFF;
+
+	if (data_in & BIT(11)) {
+		data_in |= ADXL362_COMPLEMENT;
+	}
+
+	int64_t micro_ms2 = data_in * SENSOR_G / range_to_scale[range];
+
+	*out = CLAMP((micro_ms2 + (micro_ms2 % 1000000)), INT32_MIN, INT32_MAX);
```

Comment:
the modulo and division above are going to be pretty costly I'd think? I again wonder if this can't be a simple scale multiplier to convert from what is likely already a fixed point adc reading to another scale and numerical container size

See #76968